src/jl_uv.c

```diff
@@ -65,17 +66,22 @@ void jl_init_signal_async(void)
 }
 #endif
 
-#ifdef JULIA_ENABLE_THREADING
-JL_DLLEXPORT void jl_uv_lock()
+static jl_mutex_t jl_uv_mutex;
+
+JL_DLLEXPORT void jl_uv_lock(void)
 {
-    // TODO: put recursive lock here
+    JL_LOCK_NOGC(&jl_uv_mutex);
```

Comment:
I'm 99% sure this should not be the `nogc` variance.

Reply:
ok, I trust your judgement here since I have no clue about GC.. but I just wanted to to keep the non-threaded code intact in its behavior

Reply:
```Program received signal SIGSEGV, Segmentation fault.
0x00007ffff76b5126 in jl_mutex_unlock (lock=<optimized out>) at /nfs/an/proj/scripting_tools/amalakho/projects/julia/src/locks.h:138
138         JL_SIGATOMIC_END();
(rr) bt
#0  0x00007ffff76b5126 in jl_mutex_unlock (lock=<optimized out>) at /nfs/an/proj/scripting_tools/amalakho/projects/julia/src/locks.h:138
#1  jl_atexit_hook (exitcode=exitcode@entry=1) at /nfs/an/proj/scripting_tools/amalakho/projects/julia/src/init.c:290
#2  0x00007ffff76d45a7 in jl_exit (exitcode=exitcode@entry=1) at /nfs/an/proj/scripting_tools/amalakho/projects/julia/src/jl_uv.c:645
#3  0x00007ffff770a981 in jl_vexceptionf (exception_type=<optimized out>, fmt=0x7ffff781b688 "julia: invalid argument to --compile (%s)", args=0x7fffffffd7c0)
    at /nfs/an/proj/scripting_tools/amalakho/projects/julia/src/rtutils.c:53
#4  0x00007ffff770aada in jl_errorf (fmt=fmt@entry=0x7ffff781b688 "julia: invalid argument to --compile (%s)")
    at /nfs/an/proj/scripting_tools/amalakho/projects/julia/src/rtutils.c:75
#5  0x00007ffff76fb918 in jl_parse_opts (argcp=argcp@entry=0x7fffffffd93c, argvp=argvp@entry=0x7fffffffd930)
    at /nfs/an/proj/scripting_tools/amalakho/projects/julia/src/jloptions.c:436
#6  0x00000000004014a8 in main (argc=<optimized out>, argv=<optimized out>) at /nfs/an/proj/scripting_tools/amalakho/projects/julia/ui/repl.c:226
```
Bad luck with 1%.. I'll go back to nogc option unless you think that gc is strongly preferred and can suggest how to fix the exit hook above.

Reply:
Nogc is wrong. It's not just a preference

Reply:
why? please give me a technical argument. there are tons of safepoints around already and GC-enabled lock will just duplicate these, also it just fails and changes behavior even without threads being enabled (by adding these GC points there)

Reply:
isn't it a goal? UV operations must be fast in order to avoid contention on the lock, Julia code being called in the callbacks is restricted already, e.g. one cannot yield there. Is it currently possible that there are GC safepoints inside libuv callbacks?

Reply:
Isn't what a goal? Be fast isn't a goal when it's wrong. Yielding is orders of magnitude worse than safepoint. No Julia code is allowed in the nogc critical region which I thought I stated pretty clearly in the comments.

> JL_LOCK_NOGC should only be needed for GC internal locks

> Is it currently possible that there are GC safepoints inside libuv callbacks?

Why not.

Reply:
In other words - how can I disable GC for the duration of the uv call? Is it possible with all these callbacks implemented in Julia ?

Reply:
No, disabling gc when calling Julia code is wrong. And it'll be very slow.

Reply:
> I guess when you are a developer who changes things, nothing is cut in stone to be called just wrong, it's a matter of whether there is a good argument or a better way.

Not really. There are things that are fundamental enough that can't be changed.

> I don't object to GC-enabled locks for now - just please suggest me how to fix that segfault.

I don't even know what segfault you are talking about....................... If it's just exiting, just disable something could be the right solution.

> Going forward, it will become a performance bottleneck since I believe GC might take a while to complete, so doing it inside the critical section is not a good idea.

I really don't know what you are talking about here. Using gc or no gc lock has nothing to do with doing GC in the critical region or not. The GC is a global operation that does not individually affect the performance of each thread. This should hold true even for parallel GC and could still be true for concurrent GC.

> Basically, doing anything heavy inside a critical section is generally a bad design approach, wouldn't you agree?

So I don't agree.

Reply:
ok, you are right about the GC under the lock as it has GC safepoint inside of the busy waiting loop and this GC has to stop the world anyway, thanks for explanations.